server/base/src/main/java/org/apache/accumulo/server/conf/util/ConfigTransformer.java

```diff
@@ -0,0 +1,366 @@
+
+  /**
+   * Instantiate a transformer instance.
+   *
+   * @param zrw
+   *          a ZooReaderWriter
+   * @param codec
+   *          the codec used to encode to the single-node format.
+   * @param propStoreWatcher
+   *          the watcher registered to receive future notifications of changes to the encoded
+   *          property node.
+   */
+  public ConfigTransformer(final ZooReaderWriter zrw, VersionedPropCodec codec,
+      final PropStoreWatcher propStoreWatcher) {
+    this.zrw = zrw;
+    this.codec = codec;
+    this.propStoreWatcher = propStoreWatcher;
+
+    // default - allow for a conservative max delay of about a minute
+    retry =
+        Retry.builder().maxRetries(15).retryAfter(250, MILLISECONDS).incrementBy(500, MILLISECONDS)
+            .maxWait(5, SECONDS).backOffFactor(1.75).logInterval(3, MINUTES).createRetry();
+
+  }
+
+  public ConfigTransformer(final ZooReaderWriter zrw, VersionedPropCodec codec,
+      final PropStoreWatcher propStoreWatcher, final Retry retry) {
+    this.zrw = zrw;
+    this.codec = codec;
+    this.propStoreWatcher = propStoreWatcher;
+    this.retry = retry;
+  }
+
+  /**
+   * Transform the properties for the provided prop cache key.
+   *
+   * @return the encoded properties.
+   */
+  public VersionedProperties transform(final PropCacheKey propCacheKey) {
+    TransformLock lock = TransformLock.createLock(propCacheKey, zrw);
+    return transform(propCacheKey, lock);
+  }
+
+  // Allow external (mocked) TransformLock to be used
+  @VisibleForTesting
+  VersionedProperties transform(final PropCacheKey propCacheKey, final TransformLock lock) {
+
+    VersionedProperties results;
+    Instant start = Instant.now();
+    try {
+      while (!lock.isLocked()) {
+        try {
+          retry.useRetry();
+          retry.waitForNextAttempt();
+          // look and return node if created.
+          if (zrw.exists(propCacheKey.getPath())) {
+            Stat stat = new Stat();
+            byte[] bytes = zrw.getData(propCacheKey.getPath(), propStoreWatcher, stat);
+            return codec.fromBytes(stat.getVersion(), bytes);
+          }
+          // still does not exist - try again.
+          lock.lock();
+        } catch (InterruptedException ex) {
+          Thread.currentThread().interrupt();
+          throw new PropStoreException("Failed to get transform lock for " + propCacheKey, ex);
+        } catch (IllegalStateException ex) {
+          throw new PropStoreException("Failed to get transform lock for " + propCacheKey, ex);
+        }
+      }
+
+      Set<LegacyPropNode> upgradeNodes = readLegacyProps(propCacheKey);
+
+      upgradeNodes = convertDeprecatedProps(propCacheKey, upgradeNodes);
+
+      results = writeConverted(propCacheKey, upgradeNodes);
+
+      if (results == null) {
+        throw new PropStoreException("Could not create properties for " + propCacheKey, null);
+      }
+
+      // validate lock still valid before deletion.
+      if (!lock.validateLock()) {
```

Comment:
What situation does this defend against?  The lock could be "valid" but become "not valid" after this call and before the call to `deleteLegacyProps`

Reply:
The locking was refactored in fbb86bafe9

The two-level caching was existing behavior and was necessary to support the updateCount scheme used by the derivers - because of the frequency that the updateCount is checked, I attempted to avoid locking and synchronization except during updates.  The updateCount deals with the property hierarchy - which seems expensive to compute and why the two-level caching was used in the first place.  

The refactored code in fbb86bafe9 should resolve the comments on locking and multiple calls to set the snapshot to null.

The code may have race conditions because getProperties and getDataVersion in the configuration are made in separate methods and could result in calls to different snapshots - however, I think that any races would be handled because things will be eventually consistent. Access to the property map should always be the most recent, and if a call gets an earlier data version and then reads a more recent set of properties - those properties should be "more" correct - and then on the next data version check, the count will have advanced to match the properties.  It will then re-read, re-build the configuration with the values that it already had, but they are still correct.  This should be rare and seems a reasonable tradeoff - otherwise additional synchronization / locking would be needed to make sure they are strictly consistent - and that synchronization / locking likely would impact the deriver performance. (Well, that is the goal)

Reply:
Additional improvements in removing the two-level caching can be addressed in https://github.com/apache/accumulo/issues/2634

Reply:
I get kinda lost in the code w/ all the levels and thread pools when I review.  I looked for dead locks and race conditions but was unable to do a comprehensive analysis for those and only did a partial analysis.